src/circuit/gadget/sinsemilla/chip.rs

```diff
@@ -0,0 +1,365 @@
+        &self.config
+    }
+
+    fn loaded(&self) -> &Self::Loaded {
+        &()
+    }
+}
+
+impl SinsemillaChip {
+    pub fn construct(config: <Self as Chip<pallas::Base>>::Config) -> Self {
+        Self { config }
+    }
+
+    pub fn load(
+        config: SinsemillaConfig,
+        layouter: &mut impl Layouter<pallas::Base>,
+    ) -> Result<<Self as Chip<pallas::Base>>::Loaded, Error> {
+        // Load the lookup table.
+        config.generator_table.load(layouter)
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    #[allow(non_snake_case)]
+    pub fn configure(
+        meta: &mut ConstraintSystem<pallas::Base>,
+        advices: [Column<Advice>; 5],
+        lookup: (Column<Fixed>, Column<Fixed>, Column<Fixed>),
+        constants: Column<Fixed>,
+        perm: Permutation,
+    ) -> <Self as Chip<pallas::Base>>::Config {
+        let config = SinsemillaConfig {
+            q_sinsemilla1: meta.selector(),
+            q_sinsemilla2: meta.fixed_column(),
+            fixed_y_q: meta.fixed_column(),
+            x_a: advices[0],
+            x_p: advices[1],
+            bits: advices[2],
+            lambda_1: advices[3],
+            lambda_2: advices[4],
+            generator_table: GeneratorTableConfig {
+                table_idx: lookup.0,
+                table_x: lookup.1,
+                table_y: lookup.2,
+            },
+            constants,
+            perm,
+        };
+
+        // Set up lookup argument
+        GeneratorTableConfig::configure(meta, config.clone());
+
+        // Constant expressions
+        let two = Expression::Constant(pallas::Base::from_u64(2));
+
+        // Closures for expressions that are derived multiple times
+        // x_r = lambda_1^2 - x_a - x_p
+        let x_r = |meta: &mut VirtualCells<pallas::Base>, rotation| {
+            let x_a = meta.query_advice(config.x_a, rotation);
+            let x_p = meta.query_advice(config.x_p, rotation);
+            let lambda_1 = meta.query_advice(config.lambda_1, rotation);
+            lambda_1.square() - x_a - x_p
+        };
+
+        // Y_A = (lambda_1 + lambda_2) * (x_a - x_r)
+        let Y_A = |meta: &mut VirtualCells<pallas::Base>, rotation| {
+            let x_a = meta.query_advice(config.x_a, rotation);
+            let lambda_1 = meta.query_advice(config.lambda_1, rotation);
+            let lambda_2 = meta.query_advice(config.lambda_2, rotation);
+            (lambda_1 + lambda_2) * (x_a - x_r(meta, rotation))
+        };
+
+        // Check that the initial x_A, x_P, lambda_1, lambda_2 are consistent with y_Q.
+        meta.create_gate("Initial y_Q", |meta| {
+            let fixed_y_q = meta.query_fixed(config.fixed_y_q, Rotation::cur());
+
+            // Y_A = (lambda_1 + lambda_2) * (x_a - x_r)
+            let Y_A = Y_A(meta, Rotation::cur());
+
+            // fixed_y_q * (2 * fixed_y_q - Y_{A,0}) = 0
+            vec![fixed_y_q.clone() * (two.clone() * fixed_y_q - Y_A)]
+        });
+
```

Comment:
If we move this inside the "Sinsemilla gate", then `MockProver` will see the `fixed_y_q` query as semantically connected to `q_sinsemilla1` (which it is - we only want to use this constraint specifically in the context of Sinsemilla), and ensure that this constraint is assigned correctly. It _would_ mean that we need to explicitly assign zero to every row except the first within `SinsemillaChip::hash_message`, but I think that's fine.